### PR TITLE
Reworks Bear Traps

### PR DESCRIPTION
**Pull request by Burzah**
## What Does This PR Do
This PR reworks how bear traps function. It makes them take a few seconds to arm and will trigger if they are in a bag when it is opened.  This rework also reduces the amount of time it takes to remove a trap for a leg, and applies extra damage to bears and other simple mobs that trigger the trap.

## Why It's Good For The Game
Stops people from carrying multiple loaded bear traps in a closed bag, while providing some extra QoL adjustments.

## Images of changes
dreamseeker_FMob5haGSu

## Testing
Tested to see if an armed bear trap would trigger if stored in a bag and then opened. Tested to see if the behaviors were also the same with current bear trap code that allows for signalers and IEDs to be attached to them.
Tested to see if timer was functioning properly.
Tested to see if bears took the proper amount of damage.

## Changelog
:cl:
tweak:  bear traps in bags can trigger and hurt the opener
tweak: bear traps now have a 2 second arm time
tweak: bear trap removal from a leg reduced from 30 seconds to 20 seconds
tweak: bear traps now do extra damage to simple mobs and a bit more to bears
/:cl:

### CODE REVIEW

File: code/game/objects/items/weapons/legcuffs.dm [hunk 1]
```diff
@@ -130,6 +130,31 @@
...
 
+/obj/item/restraints/legcuffs/beartrap/on_found(mob/finder)
+	if(!armed)
+		return FALSE
+	var/mob/living/L = finder
+	armed = FALSE
+	update_icon()
+	playsound(src.loc, 'sound/effects/snap.ogg', 50, 1)
+	if(!silent_arming)
+		L.visible_message("<span class='danger'>[L] triggers \the [src].</span>", "<span class='userdanger'>You trigger \the [src]!</span>")
```
Comment (DGamerL) [inline code/game/objects/items/weapons/legcuffs.dm:141]:
```suggestion
		L.visible_message("<span class='danger'>[L] triggers [src].</span>", "<span class='userdanger'>You trigger [src]!</span>")
```
\the is implicit, no need to have these (Yeah I know it's not your code but still)

File: code/game/objects/items/weapons/legcuffs.dm [hunk 2]
```diff
@@ -41,7 +41,11 @@
 
 /obj/item/restraints/legcuffs/beartrap/attack_self(mob/user)
 	..()
-	if(ishuman(user) && !user.stat && !user.restrained())
+
+	if(!ishuman(user) || ishuman(user) && user.restrained())
+		return
+
+	if(do_after(user, 2 SECONDS, TRUE, user, TRUE))
```
Comment (DGamerL) [inline code/game/objects/items/weapons/legcuffs.dm:48]:
```suggestion
	if(do_after(user, 2 SECONDS, target = user))
```
These values are already `TRUE` by default, no need to redeclare them

File: code/game/objects/items/weapons/legcuffs.dm [hunk 3]
```diff
@@ -41,7 +41,11 @@
 
 /obj/item/restraints/legcuffs/beartrap/attack_self(mob/user)
 	..()
-	if(ishuman(user) && !user.stat && !user.restrained())
+
+	if(!ishuman(user) || ishuman(user) && user.restrained())
+		return
+
+	if(do_after(user, 2 SECONDS, target = user))
```
Comment (DGamerL) [inline code/game/objects/items/weapons/legcuffs.dm:48]:
```suggestion
	if(do_after(user, 2 SECONDS))
```
My bad this also doesn't work. Was there any goal you tried with the target being on the user?

Comment (Burzah) [reply code/game/objects/items/weapons/legcuffs.dm:48]: Talked about this over discord with each other and resolved the issue.

File: code/game/objects/items/weapons/legcuffs.dm [hunk 4]
```diff
@@ -127,9 +132,37 @@
...
+					L.apply_damage(trap_damage, BRUTE)
 	..()
 
+/obj/item/restraints/legcuffs/beartrap/on_found(mob/finder)
+	if(!armed)
+		return FALSE
+	var/mob/living/L = finder
+	armed = FALSE
+	update_icon()
+	playsound(src.loc, 'sound/effects/snap.ogg', 50, 1)
```
Comment (Henri215) [inline code/game/objects/items/weapons/legcuffs.dm:147]:
```suggestion
	playsound(loc, 'sound/effects/snap.ogg', 50, TRUE)
```

File: code/game/objects/items/weapons/legcuffs.dm [hunk 5]
```diff
@@ -41,7 +42,11 @@
 
 /obj/item/restraints/legcuffs/beartrap/attack_self(mob/user)
 	..()
-	if(ishuman(user) && !user.stat && !user.restrained())
+
+	if(!ishuman(user) || ishuman(user) && user.restrained())
```
Comment (Henri215) [inline code/game/objects/items/weapons/legcuffs.dm:46]: You can avoid checking for "ishuman()" a second time here.

File: code/game/objects/items/weapons/legcuffs.dm [hunk 6]
```diff
@@ -127,9 +132,37 @@
...
+					L.apply_damage(trap_damage * 2.5, BRUTE)
+				else
+					L.apply_damage(trap_damage, BRUTE)
 	..()
 
+/obj/item/restraints/legcuffs/beartrap/on_found(mob/finder)
+	if(!armed)
+		return FALSE
+	var/mob/living/L = finder
+	armed = FALSE
```
Comment (hal9000PR) [inline code/game/objects/items/weapons/legcuffs.dm:145]: lots of copy pasted code here from `crossed` make a generic proc called `spring_trap` or something

**Review by hal9000PR — CHANGES_REQUESTED** [review]

File: code/game/objects/items/weapons/legcuffs.dm (hunk 5 above)
Comment (Burzah) [reply code/game/objects/items/weapons/legcuffs.dm]: resolved with [7383b94](https://github.com/ParadiseSS13/Paradise/pull/21941/commits/7383b9491f7e56b5739fa51c130edd7c766e0de5)

File: code/game/objects/items/weapons/legcuffs.dm [hunk 7]
```diff
@@ -127,9 +134,32 @@
...
+
+	if(IED)
+		IED.active = TRUE
+		message_admins("[key_name_admin(usr)] has triggered an IED-rigged [name].")
+		log_game("[key_name(usr)] has triggered an IED-rigged [name].")
+		spawn(IED.det_time)
+		IED.prime()
+
+	if(sig)
+		sig.signal()
```
Comment (DGamerL) [inline code/game/objects/items/weapons/legcuffs.dm:156]: Same for the `if(sig)` checks

File: code/game/objects/items/weapons/legcuffs.dm [hunk 8]
```diff
@@ -93,16 +98,18 @@
 		to_chat(user, "<span class='notice'>You remove the signaler from [src].</span>")
 	return TRUE
 
+/obj/item/restraints/legcuffs/beartrap/proc/spring_trap(mob/user)
+	armed = FALSE
+	update_icon()
+	playsound(loc, 'sound/effects/snap.ogg', 50, TRUE)
+	if(!silent_arming)
+		user.visible_message("<span class='danger'>[user] triggers [src].", "<span class='userdanger'>You trigger [src].")
+
```
Comment (DGamerL) [inline code/game/objects/items/weapons/legcuffs.dm:107]: Could you move this code to below where it's used?

File: code/game/objects/items/weapons/legcuffs.dm [hunk 9]
```diff
@@ -98,16 +98,18 @@
...
-			armed = FALSE
-			update_icon()
-			playsound(src.loc, 'sound/effects/snap.ogg', 50, 1)
-			if(!silent_arming)
-				L.visible_message("<span class='danger'>[L] triggers \the [src].</span>", \
-						"<span class='userdanger'>You trigger \the [src]!</span>")
+			spring_trap(AM)
 
 			if(IED && isturf(src.loc))
 				IED.active = TRUE
```
[the remark this answers is not in the document]
Comment (Burzah) [reply code/game/objects/items/weapons/legcuffs.dm]: Resolved with [eff2394](https://github.com/ParadiseSS13/Paradise/pull/21941/commits/eff23948f988a95efe39da1b0bbc341ba521e50c) as well as the the signaler check.

File: code/game/objects/items/weapons/legcuffs.dm (hunk 6 above)
Comment (Burzah) [reply code/game/objects/items/weapons/legcuffs.dm]: Should be resolved with [eff2394](https://github.com/ParadiseSS13/Paradise/pull/21941/commits/eff23948f988a95efe39da1b0bbc341ba521e50c)

File: code/game/objects/items/weapons/legcuffs.dm (hunk 8 above)
Comment (Burzah) [reply code/game/objects/items/weapons/legcuffs.dm]: Resolved in [ce20a72](https://github.com/ParadiseSS13/Paradise/pull/21941/commits/ce20a72729e1183110f0a9ebde2fd060fde8b16c)

**Comment by Burzah** [comment]
Added in a damage modifier for simple mobs that is less than the damage mod of bears.

File: code/game/objects/items/weapons/legcuffs.dm [hunk 10]
```diff
@@ -127,9 +120,39 @@
...
+				if(istype(L, /mob/living/simple_animal/hostile/bear))
+					L.apply_damage(trap_damage * 2.5, BRUTE)
+				else
+					L.apply_damage(trap_damage * 1.75, BRUTE)
 	..()
 
+/obj/item/restraints/legcuffs/beartrap/on_found(mob/finder)
+	if(!armed)
+		return FALSE
+	spring_trap(finder, IED, sig)
```
Comment (hal9000PR) [inline code/game/objects/items/weapons/legcuffs.dm:132]: why is IED a parameter when its a var stored on the beartrap. same with signaller.

**Review by hal9000PR — CHANGES_REQUESTED** [review]

**Review by hal9000PR — APPROVED** [review]

File: code/game/objects/items/weapons/legcuffs.dm [hunk 11]
```diff
@@ -94,42 +99,59 @@
...
+		var/mob/living/L = AM
+		spring_trap(AM)
 
+		if(ishuman(AM))
+			var/mob/living/carbon/H = AM
+			if(IS_HORIZONTAL(H))
+				H.apply_damage(trap_damage, BRUTE,"chest")
 			else
-				L.apply_damage(trap_damage, BRUTE)
+				H.apply_damage(trap_damage, BRUTE,(pick("l_leg", "r_leg")))
```
Comment (Henri215) [inline code/game/objects/items/weapons/legcuffs.dm:114]:
```suggestion
				H.apply_damage(trap_damage, BRUTE, pick("l_leg", "r_leg"))
```

File: code/game/objects/items/weapons/legcuffs.dm [hunk 12]
```diff
@@ -94,42 +99,58 @@
...
-					H.apply_damage(trap_damage, BRUTE,(pick("l_leg", "r_leg")))
-				if(!H.legcuffed && H.get_num_legs() >= 2) //beartrap can't cuff you leg if there's already a beartrap or legcuffs.
-					H.legcuffed = src
-					forceMove(H)
-					H.update_inv_legcuffed()
-					SSblackbox.record_feedback("tally", "handcuffs", 1, type)
+	if(!armed || !isturf(loc))
+		return ..()
+
+	if((iscarbon(AM) || isanimal(AM)) && !istype(AM, /mob/living/simple_animal/parrot) && !isconstruct(AM) && !isshade(AM) && !istype(AM, /mob/living/simple_animal/hostile/viscerator))
```
Comment (DGamerL) [inline code/game/objects/items/weapons/legcuffs.dm:105]:
```suggestion
	if((iscarbon(AM) || isanimal(AM)) && AM.flying != TRUE))
```
Wouldn't this be exactly what Henri asked?

Comment (Burzah) [reply code/game/objects/items/weapons/legcuffs.dm]: Adjusted in [81a2414](https://github.com/ParadiseSS13/Paradise/pull/21941/commits/81a241446d343b2d0b9bae1efa0b926e51dc2fa5)

File: code/game/objects/items/weapons/legcuffs.dm [hunk 13]
```diff
@@ -94,42 +99,58 @@
...
-				if(!H.legcuffed && H.get_num_legs() >= 2) //beartrap can't cuff you leg if there's already a beartrap or legcuffs.
-					H.legcuffed = src
-					forceMove(H)
-					H.update_inv_legcuffed()
-					SSblackbox.record_feedback("tally", "handcuffs", 1, type)
+	if(!armed || !isturf(loc))
+		return ..()
+
+	var/mob/living/L = AM
+	if((iscarbon(AM) || isanimal(AM)) && L.flying != TRUE)
```
Comment (Henri215) [inline code/game/objects/items/weapons/legcuffs.dm:106]:
```suggestion
	if((iscarbon(AM) || isanimal(AM)) && !L.flying)
```

File: code/game/objects/items/weapons/legcuffs.dm [hunk 14]
```diff
@@ -94,42 +99,58 @@
...
+		var/mob/living/carbon/H = finder
+		H.apply_damage(trap_damage, BRUTE, pick("l_hand", "r_hand"))
+	return TRUE
+
+/obj/item/restraints/legcuffs/beartrap/proc/spring_trap(mob/user)
+	armed = FALSE
+	update_icon()
+	playsound(loc, 'sound/effects/snap.ogg', 50, TRUE)
+	if(!silent_arming)
+		user.visible_message("<span class='danger'>[user] triggers [src].", "<span class='userdanger'>You trigger [src].")
```
Comment (Henri215) [inline code/game/objects/items/weapons/legcuffs.dm:142]:
```suggestion
		user.visible_message("<span class='danger'>[user] triggers [src].</span>", "<span class='userdanger'>You trigger [src].</span>")
```

**Review by Henri215 — APPROVED** [review]
LGTM

**Review by DGamerL — APPROVED** [review]

**Pull request merged**